src/freenas/usr/local/bin/truenas-initrd.py

```diff
@@ -14,38 +15,95 @@
 
 logger = logging.getLogger(__name__)
 
-readonly_state = None
 
+@dataclasses.dataclass
+class ReadonlyState:
+    original_root: bool = True
+    original_usr: bool = True
+    current_root: bool = True
+    current_usr: bool = True
 
-def set_readonly(root, readonly):
+
+readonly_state = ReadonlyState()
+
+
+def initialize_readonly_state(root: str):
+    global readonly_state
+
+    mountpoints = {
+        root: "root",
+        os.path.join(root, "usr"): "usr",
+    }
+
+    for partition in psutil.disk_partitions():
```

Comment:
Do we really need to iterate over every mounted filesystem just to query a very select few datasets? I'm wondering if we just query the datasets recursively?

Finally, if we were to use `getmntinfo` function, we can parse the mount options for `RO` instead of having to subprocess out inside a loop.

Reply:
@anodos325 any input on this loop? I believe we can get around having to subprocess entirely but maybe I'm missing something obvious.

Reply:
When TrueNAS image is created we create a file `/conf/truenas_root_ds.json` that includes all the dataset names as well as very initial options that we set. Perhaps we can just read that file and select the datasets we care about.